File: 00Playground/adventOfCode/Day11_15/Day15.py
import re

def main():
    fin = open("Day15.in", "r")
    coor = []
    ans = 0
    inps = []
    for i in range(23):
        inp = re.split("=|:|,", fin.readline().strip())
        # print(inp)
        inps.append([i for i in [inp[1], inp[3], inp[5], inp[7]]])
        # print(" ".join(inps[-1]))
    for j in range(0, 200000):
        for i in range(23):
            curr = inps[i]
            # print(curr)
            manhattan_dist = abs(curr[2] - curr[0]) + abs(curr[3] - curr[1])
            start = curr[0] - (manhattan_dist - abs(j - curr[1]))
            end = curr[0] + (manhattan_dist - abs(j - curr[1]))
            # print(start, end)
            if start <= end:
                coor.append([start, end])
        final = []
        coor.sort()
        curr = coor[0]
        for i in range(1, len(coor)):
            if curr[1] >= coor[i][0]: curr[1] = max(curr[1], coor[i][1])
            else:
                final.append(curr)
                curr = coor[i]
        final.append(curr)
        if len(final) != 1: print(final, j)
        # print(final)


if __name__ == '__main__':
    main()

File: 00Playground/adventOfCode/Day11_15/Day15.cpp
#include <algorithm>
#include <fstream>
#include <iostream>
#include <vector>

using namespace std;

ifstream fin("Day15.in");

int main() {
  vector<vector<int>> coor;
  for (int i = 0; i < 23; i++) {
    vector<int> ne;
    for (int i = 0; i < 4; i++) {
      int a;
      fin >> a;
      // cout << a << ' ';
      ne.push_back(a);
    }
    // cout << endl;
    coor.push_back(ne);
  }
  int i;
  for (i = 0; i < 4000000; i++) {
    vector<pair<int, int>> covered;
    for (int j = 0; j < 23; j++) {
      vector<int> curr = coor[j];
      int manhattan_dist = abs(curr[2] - curr[0]) + abs(curr[3] - curr[1]);
      int start = curr[0] - (manhattan_dist - abs(i - curr[1]));
      int end = curr[0] + (manhattan_dist - abs(i - curr[1]));
      if (start <= end)
        covered.push_back(make_pair(start, end));
    }
    sort(covered.begin(), covered.end());
    int end = covered[0].first;
    for (pair<int, int> x : covered) {
      if (end < x.first) {
        cout << i << ' ' << end << ' ' << x.first << endl;
        return 0;
      }
      if (end > 4000000) break;
      end = max(end, x.second);
      // cout << x.first << ' ' << x.second << ' ' << end << endl;
    }
  }
  cout << i << endl;
}
